dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb

```diff
@@ -0,0 +1,75 @@
+shared_examples_for 'Creates a dragonfly App:' do
+
+  describe 'app creation' do
+    ::Refinery::Dragonfly.configure!(described_class)
+    it 'exists' do
+      expect(::Dragonfly.app(described_class.dragonfly_name).name).to eq(described_class.dragonfly_name.to_sym)
+    end
+  end
+
+  describe 'app configuration' do
+    # These are just some of the configuration parameters which can be passed to Dragonfly
+    it 'sets the datastore_root_path' do
+      expect(Dragonfly.app(described_class.dragonfly_name).datastore.root_path).to eq(described_class.dragonfly_datastore_root_path)
+    end
+
+    it 'sets the url host' do
+      expect(Dragonfly.app(described_class.dragonfly_name).server.url_host).to eq(described_class.dragonfly_url_host)
+    end
+
+    it 'sets the url format' do
+      expect(Dragonfly.app(described_class.dragonfly_name).server.url_format).to eq(described_class.dragonfly_url_format)
+    end
+  end
+
+  describe 'configuring Amazon S3' do
+    context "using custom settings" do
+      it "uses the custom values" do
+        Refinery::Dragonfly.s3_bucket_name = "kfc"
+        described_class.s3_bucket_name = "buckethead"
+
+        expect(described_class.s3_bucket_name).to eq("buckethead")
+      end
+    end
+
+    describe "without custom settings" do
+      it "uses the defaults" do
+        described_class.s3_bucket_name = nil
+        Refinery::Dragonfly.s3_bucket_name = "kfc"
+
+        expect(described_class.s3_bucket_name).to eq("kfc")
+      end
+    end
+  end
+
+  describe "defining a custom datastore" do
+    before(:all) do
+      CustomDatastore1 = Class.new
+      CustomDatastore2 = Class.new
```

Comment:
this throws up warnings because these are shared examples so when they get redefined we see a warning from Ruby:

```
refinerycms/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb:47: warning: already initialized constant CustomDatastore1
refinerycms/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb:47: warning: previous definition of CustomDatastore1 was here
refinerycms/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb:48: warning: already initialized constant CustomDatastore2
refinerycms/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb:48: warning: previous definition of CustomDatastore2 was here
```

Do they need to be constants?

Reply:
Or, in the case that they do need to be, can they be redefined in a way that avoids the warning? 😄

Reply:
In fact, it can be done like this:

```diff
diff --git a/dragonfly/spec/support/custom_datastore_one.rb b/dragonfly/spec/support/custom_datastore_one.rb
new file mode 100644
index 000000000..30871cf4c
--- /dev/null
+++ b/dragonfly/spec/support/custom_datastore_one.rb
@@ -0,0 +1 @@
+CustomDatastoreOne = Class.new
diff --git a/dragonfly/spec/support/custom_datastore_two.rb b/dragonfly/spec/support/custom_datastore_two.rb
new file mode 100644
index 000000000..7307b260b
--- /dev/null
+++ b/dragonfly/spec/support/custom_datastore_two.rb
@@ -0,0 +1 @@
+CustomDatastoreTwo = Class.new
diff --git a/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb b/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb
index f22873cb9..bb13dc27f 100644
--- a/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb
+++ b/dragonfly/spec/support/shared_examples/creates_a_dragonfly_app.rb
@@ -1,5 +1,7 @@
-shared_examples_for 'Creates a dragonfly App:' do
+require_relative '../custom_datastore_one'
+require_relative '../custom_datastore_two'

+shared_examples_for 'Creates a dragonfly App:' do
   describe 'app creation' do
     ::Refinery::Dragonfly.configure!(described_class)
     it 'exists' do
@@ -43,13 +45,8 @@ shared_examples_for 'Creates a dragonfly App:' do
   end

   describe "defining a custom datastore" do
-    before(:all) do
-      CustomDatastore1 = Class.new
-      CustomDatastore2 = Class.new
-    end
-
     before(:each) do
-      allow(Refinery::Dragonfly).to receive_messages(custom_datastore_class: CustomDatastore1)
+      allow(Refinery::Dragonfly).to receive_messages(custom_datastore_class: CustomDatastoreOne)
     end

     after(:each) do
@@ -58,14 +55,14 @@ shared_examples_for 'Creates a dragonfly App:' do

     context "with the default configuration" do
       it "uses the default values" do
-        expect(described_class.dragonfly_custom_datastore_class).to eq(CustomDatastore1)
+        expect(described_class.dragonfly_custom_datastore_class).to eq(CustomDatastoreOne)
       end
     end

     context "with a custom configuration" do
       it "uses custom values" do
-        described_class.dragonfly_custom_datastore_class = 'CustomDatastore2'
-        expect(described_class.dragonfly_custom_datastore_class).to eq(CustomDatastore2)
+        described_class.dragonfly_custom_datastore_class = 'CustomDatastoreTwo'
+        expect(described_class.dragonfly_custom_datastore_class).to eq(CustomDatastoreTwo)
       end
     end
```

Reply:
👍

Reply:
What about to use `let` here?

Reply:
@simi yeah, I thought about that too, but it leads to an issue where the class name doesn't exist and so it can't resolve it. We could mock that point out, though.